ibis/expr/tests/test_schema.py

```diff
@@ -393,6 +393,13 @@ def test_schema_from_to_pyarrow_schema():
     assert restored_schema == pyarrow_schema
 
 
+def test_schema___arrow_c_schema__():
+    pytest.importorskip("pyarrow")
+    schema = sch.Schema({"a": dt.int64, "b": dt.string, "c": dt.boolean})
+    # smoketest, since no way to create schema from capsule in current pyarrow
+    assert schema.__arrow_c_schema__() is not None
```

Comment:
`pa.schema` will use the `__arrow_c_schema__` method for recent pyarrow (v15+ I think).

I sometimes use a wrapper class in tests to ensure it sees only the `__arrow_c_schema__` method:
```py
class CSchemaWrapper:
    def __init__(obj):
        self.obj = obj

    def __arrow_c_schema__(self):
        return self.obj.__arrow_c_schema__()
```
then check
```py
pa.schema(CSchemaWrapper(schema))
```

Reply:
fwiw, I tried this (`pa.schema(ibis_schema)`) with the ibis 9.2.0 and alas, no dice:

```console
File [~/test-hook/lib/python3.11/site-packages/pyarrow/types.pxi:5360](https://notebooks.freenome.net/user/mcorley/lab/tree/test-hook/lib/python3.11/site-packages/pyarrow/types.pxi#line=5359), in pyarrow.lib.schema()

File [~/test-hook/lib/python3.11/site-packages/pyarrow/types.pxi:3541](https://notebooks.freenome.net/user/mcorley/lab/tree/test-hook/lib/python3.11/site-packages/pyarrow/types.pxi#line=3540), in pyarrow.lib.field()

File [~/test-hook/lib/python3.11/site-packages/pyarrow/types.pxi:5301](https://notebooks.freenome.net/user/mcorley/lab/tree/test-hook/lib/python3.11/site-packages/pyarrow/types.pxi#line=5300), in pyarrow.lib.ensure_type()

TypeError: DataType expected, got <class 'ibis.expr.datatypes.core.String'>

Reply:
Arg, pyarrow **first** checks whether the input is a `Mapping` and then checks whether `__arrow_c_schema__` exists 😞

Reply:
Seems like this functionality is broken in pyarrow: an object can't be both a mapping *and* implement `__arrow_c_schema__` interface.

Reply:
Opened up https://github.com/apache/arrow/issues/43388 to track this.

Reply:
In the meantime I'll add an `xfail`ing test that will start to pass once this is addressed.

Reply:
#9671